Add steps for ISOMIP+ base and culled meshes

xylar opened the pull request:
<!--
Thank you for your pull request.
Please add a description of what is accomplished in the PR here at the top:
-->

This PR represents the first step in porting the ISOMIP+ experiments over from Compass (or, in the case of Ocean 3 and 4, adding new experiments).

I have opted for 3 resolutions here: 1, 2 and 4 km.  We had thought long ago that 5 km might be a "typical" resolution for Antarctic cavities in MPAS-Ocean but obviously 12 km is our standard and that is too coarse for ISOMIP+.  I also better want to demonstrate and test the flexibility with resolution.

For now:
* I only have a standard `z-star` version of each test case since the culled mesh will be the same in any case.
* I have both planar and spherical versions of all test cases.
* I am processing the geometry for all 5 experiments: ocean0-4
* I have retained various parameters to the constructor of `IsomipPlusTest` as placeholders as these features get ported.

There are shared steps at each resolution for creating the base mesh, putting the topography (from `ocean1`), making a mapping file from the ISOMIP+ data grid to the MPAS base mesh, remapping the topography onto the base mesh (to get the land boundary), culling the mesh, making a mapping file to the culled mesh, and remapping the topography onto the culled mesh.  After creating the planar or spherical base mesh and adding the lat/lon coordinate (to planar) or the x/y coordinate (to spherical), the following steps are agnostic to whether the mesh is planar or spherical.  

For the `insception`, `wetting` and `drying` test cases, there is a `scale` step that computes the scaled land-ice pressure and ice draft.  This will allow the modified geometry for these 3 experiments to be treated in the same way as the prescribed geometry for the `ocean3` and `ocean4` experiments.

The next step will be to add a step for creating initial conditions.

I have not yet started adding documentation because the tasks are not complete enough yet in my view to justify it.  But I could get started if folks would prefer that.

<!--
Below are a few things we ask you or your reviewers to kindly check. 
***Remove checks that are not relevant by deleting the line(s) below.***
-->
Checklist
* [ ] ~~User's Guide has been updated~~
* [ ] ~~Developer's Guide has been updated~~
* [ ] ~~API documentation in the Developer's Guide (`api.md`) has any new or modified class, method and/or functions listed~~
* [ ] ~~Documentation has been [built locally](https://e3sm-project.github.io/polaris/main/developers_guide/building_docs.html) and changes look as expected~~
* [x] `Testing` comment in the PR documents testing used to verify the changes
* [ ] ~~New tests have been added to a test suite~~

<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

xylar commented:
## Testing

I ran all 30 tasks on Chrysalis.  I saw some of the slowdown that I had been seeing in my testing of other large suites an my job died before 1 hour.  I had to delete a corrupted file at that point but after that it ran successfully through the remaining tests.

I am not inclined to investigate the slowdown right now because these large suites are not how we plan to run typically.  Either we will have task parallelism (soon I hope) or we will run smaller suites.

xylar commented:
@cbegeman, no rush at all but this is ready to review when you have the time and energy.

I plan to build on it with an `init` step in the coming week.

xylar commented:
One more comment: I think I will organize some of the steps into a `mesh` module inside of `isomip_plus` to keep things tidy.

xylar commented on polaris/ocean/tasks/isomip_plus/projection.py:
```diff
@@ -0,0 +1,39 @@
+    # like -y
+    #
+    # x_0 is close to lat=-75 and lon=90 so latitude is approximately like x
+    proj_str = ('+proj=stere +lon_0=0 +lat_0=-90 +lat_ts=-75.0 +x_0=-2000e3 '
+                '+y_0=40.0e3 +ellps=WGS84')
```
After thinking about the projection today, I decided I wasn't happy with the previous one (centered at lon=0, lat=-75) because it doesn't line up well with the x-y coordinates and it will be hard to make transects of.  This new coordinate leads to a `lonCell` that is symmetric about the middle of the domain and looks something like -y (varying mostly in the y direction), and a `latCell` that starts at about -75 degrees and looks something like x (varying mostly in the x direction).

xylar commented:
@cbegeman, based on things I'm discovering in making the `init` step(s), this should be put in draft mode and you should hold off on reviewing it for now.

xylar commented:
@cbegeman, this is ready for review.  That can be as involved or as cursory as you see fit.  Once this is in, I believe the initial condition PR is also ready to be reviewed.  I will continue on, making a forward and a viz step for configurations that will work without wetting-and-drying.

cbegeman commented:
@xylar Did you maybe forget to add `isomip_plus_topo.cfg`? I think `inception`, `wetting`, and `drying` rely on it.

```
  * step: topo_scale

polaris calling: polaris.ocean.tasks.isomip_plus.topo.scale.TopoScale.constrain_resources()
  inherited from: polaris.step.Step.constrain_resources()
  in /gpfs/fs1/home/ac.cbegeman/polaris-repo/add-isomip-plus-base-and-culled-meshes/polaris/step.py


polaris calling: polaris.ocean.tasks.isomip_plus.topo.scale.TopoScale.runtime_setup()
  inherited from: polaris.step.Step.runtime_setup()
  in /gpfs/fs1/home/ac.cbegeman/polaris-repo/add-isomip-plus-base-and-culled-meshes/polaris/step.py


polaris calling: polaris.ocean.tasks.isomip_plus.topo.scale.TopoScale.run()
  in /gpfs/fs1/home/ac.cbegeman/polaris-repo/add-isomip-plus-base-and-culled-meshes/polaris/ocean/tasks/isomip_plus/topo/scale.py

          execution:        ^[[91mERROR^[[0m
Exception raised while running the steps of the task
Traceback (most recent call last):
  File "/gpfs/fs1/home/ac.cbegeman/polaris-repo/add-isomip-plus-base-and-culled-meshes/polaris/run/serial.py", line 324, in _log_and_run_task
    baselines_passed = _run_task(task, available_resources)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/fs1/home/ac.cbegeman/polaris-repo/add-isomip-plus-base-and-culled-meshes/polaris/run/serial.py", line 403, in _run_task
    _run_step(task, step, task.new_step_log_file,
  File "/gpfs/fs1/home/ac.cbegeman/polaris-repo/add-isomip-plus-base-and-culled-meshes/polaris/run/serial.py", line 501, in _run_step
    step.run()
  File "/gpfs/fs1/home/ac.cbegeman/polaris-repo/add-isomip-plus-base-and-culled-meshes/polaris/ocean/tasks/isomip_plus/topo/scale.py", line 61, in run
    dates: List[str] = config.getlist('isomip_plus_scaling',
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ac.cbegeman/miniconda3/envs/polaris_isomipmesh/lib/python3.11/site-packages/mpas_tools/config.py", line 200, in getlist
    values = self.get(section, option)
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ac.cbegeman/miniconda3/envs/polaris_isomipmesh/lib/python3.11/site-packages/mpas_tools/config.py", line 115, in get
    return self.combined.get(section, option)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ac.cbegeman/miniconda3/envs/polaris_isomipmesh/lib/python3.11/configparser.py", line 797, in get
    d = self._unify_values(section, vars)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ac.cbegeman/miniconda3/envs/polaris_isomipmesh/lib/python3.11/configparser.py", line 1168, in _unify_values
    raise NoSectionError(section) from None
configparser.NoSectionError: No section: 'isomip_plus_scaling'
```

cbegeman commented:
> @xylar Did you maybe forget to add isomip_plus_topo.cfg? I think inception, wetting, and drying rely on it.

Or maybe it has to do with the fact that those sections are in `ocean/planar/isomip_plus/4km/isomip_plus_topo.cfg` but not in `ocean/planar/isomip_plus/4km/z-star/inception/inception.cfg` and you're trying to retrieve it from the wrong config file?

cbegeman commented on polaris/ocean/tasks/isomip_plus/mesh/cull.py:
```diff
@@ -0,0 +1,119 @@
+            make_graph_file(mesh_filename='culled_mesh.nc',
+                            graph_filename='culled_graph.info')
+
+            extract_vtk(ignore_time=True, dimension_list=['maxEdges='],
+                        variable_list=['allOnCells'],
+                        filename_pattern='culled_mesh.nc',
+                        out_dir='culled_mesh_vtk',
+                        use_progress_bar=use_progress_bar)
```
Can this be optional and turned on by a config option?

xylar replied:
I think that would be manual enough that we should just take it out.  It's not that much more involved to run the extractor command-line tool than it would be to change a config option and rerun the step.  I'm fine taking it out. It's mostly useful in the early stages of debugging.

cbegeman replied:
Ok. I'm in favor of taking it out to keep things parsimonious but I don't feel strongly about it.

xylar replied:
Done.

cbegeman commented on polaris/ocean/tasks/isomip_plus/mesh/cull.py:
```diff
@@ -0,0 +1,119 @@
+    ocean_frac = ds_topo.oceanFraction
+
+    # we want the mask to be 1 where there's not ocean
+    mask = np.logical_or(ocean_frac < 0.5, ocean_frac.isnull())
```
I don't think we need to make the min ocean fraction a config option, but can we make sure it's in the docs?

xylar replied:
It seems like the best way to document it is to make it a config option. I'll do that.

cbegeman replied:
Thanks!

xylar replied:
I added a config option.

cbegeman commented on polaris/ocean/tasks/isomip_plus/topo/remap.py:
```diff
@@ -0,0 +1,140 @@
+                                  'groundedMask': 'landIceGroundedFraction',
+                                  'openOceanMask': 'openOceanFraction',
+                                  'lowerSurface': 'landIceDraft'})
+            ds_in['oceanFraction'] = \
+                (ds_in.bedrockTopography < 0.).astype(float)
```
I remember that this is what we have used for the thin film cases. Can you point out where the grounded land ice part of the domain gets culled out when there is not a thin film? Or does that not happen?

xylar replied:
It will get handled in the initial condition by setting `maxLevelCell` to 0, so proper culling doesn't happen.

cbegeman replied:
That sounds good to me. Thanks for clarifying!

xylar replied:
I don't believe there is a to-do item here so I'm going to resolve.

cbegeman commented on polaris/ocean/tasks/isomip_plus/topo/remap.py:
```diff
@@ -0,0 +1,140 @@
+            drop = ['x', 'y', 'area', 'lat_vertices', 'lon_vertices']
+            rename = {'ncol': 'nCells'}
+            if 't' in ds_out.dims:
+                drop.append('t')
```
Why drop the variable if you are going to rename it?

xylar replied:
I believe the `t` coordinate is being dropped and the `t` dimension is being renamed to `Time` but I don't honestly recall and I wish I'd made a comment here.  I'll add one indicating that that's what I *think* I'm doing there.

cbegeman replied:
Oh, that would make some sense.

xylar replied:
I added a clarifying comment.

xylar commented:
> Did you maybe forget to add `isomip_plus_topo.cfg`?

I'll get to the bottom of this as soon as I can.  I must have a commit in my branch for adding `init` that fixes this and it should be on this branch instead.  Because my `init` branch works fine.

cbegeman commented on polaris/ocean/tasks/isomip_plus/topo/remap.py:
```diff
@@ -0,0 +1,140 @@
+                ds_out = ds_out.chunk({'Time': 1})
+            # renormalize all variables by ocean_frac
+            ocean_frac = ds_out.oceanFraction
+            for var in ds_out:
```
Is it really the case that we want to renormalize all the variables in `topography_rename.nc`? I haven't thought about this in detail, but maybe you have already? I have these variables:

```
    variables(dimensions): float64 bedrockTopography(nCells), float64 iceThickness(nCells), float64 landIceDraft(nCells), float64 landIceFloatingFraction(nCells), float64 landIceFraction(nCells), float64 landIceGroundedFraction(nCells), float64 landIcePressure(nCells), float64 oceanFraction(nCells), float64 openOceanFraction(nCells), float64 upperSurface(nCells)
```

In looking at this, I wonder whether we should have dropped `upperSurface`?

xylar replied:
Yes, I think all variables have to be renormalized to be sensible, even upper surface.  Otherwise, they drop off toward zero in places where ocean fraction is less than 1.  I don't think we currently use upper surface so I'm fine with dropping it.

xylar replied:
I removed `upperSurface`.

xylar commented:
> > Did you maybe forget to add `isomip_plus_topo.cfg`?

> I'll get to the bottom of this as soon as I can.  I must have a commit in my branch for adding `init` that fixes this and it should be on this branch instead.  Because my `init` branch works fine.

Hmm, I don't understand why my `init` branch is working because there's definitely something wrong with `scale`.  The `config` parameter is getting passed in but nothing is being done with it.  The fix should be easy enough.

xylar commented:
Hmm, ignore my last comment.  It's not that.  More debugging to do...

xylar commented:
@cbegeman, I believe I addressed your comments and fixed the issue with config files and config options (in #157).  I need to retest with the last few commits but I don't expect any surprises since they are just minor clean-up.  I'll post when my testing is done but I also think this is ready to re-review, perhaps after or along with #157.

xylar commented:
## Retesting

I ran all the ISOMIP+ test cases from this PR again on Chrysalis.  Other than a mysterious hang that I wasn't able to reproduce, everything ran fine.

xylar commented:
Rebased following the merge of #157 to hopefully make review a little simpler.

cbegeman reviewed (approved):
Thanks for responding to my comments! I reran all of the 4km cases and they pass on chyrs with intel, openmpi. I'm happy to look over the docs when you add them, in either this PR or a later one.

xylar commented:
Thanks @cbegeman.  I continue to think documentation will make more sense in a later PR so I'm going to merge this now.  I really appreciate you giving this such a thorough review.  I'm excited for the next steps!
